### DISCUSSION TITLE
Cannot let Intellisense work properly with workspace with outer submodule

### DISCUSSION THREAD

**Discussion by epiciskandar** [microsoft/vscode-cmake-tools#2580] [Q&A]
My whole project structure is like this:
(simplified)

```
./PROJ
  |--SubModuleA/
  |           |--include/<subfolders>/*.h
  |           |--src/*.cc
  |           |--CMakeLists.txt
  |--SubModuleB/
  |           |--include/<subfolders>/*.h
  |           |--src/*.cc
  |           |--CMakeLists.txt
  |--MainModule/                       <----VSCode opens at here
              |--.vscode/*.json
              |--include/<subfolders>/*.h
              |--src/*.cc
              |--CMakeLists.txt
```

Configurations:

`MainModule/.vscode/c_cpp_properties.json` :

```json
{
  "configurations": [
  {
    "configurationProvider": "ms-vscode.cmake-tools",
    "includePath": [
      "${workspaceFolder}/include",
      "${workspaceFolder}/include/**",
      "${workspaceFolder}/../SubModuleA/include/**",
      "${workspaceFolder}/../SubModuleB/include/**"
    ]
  }
  ]
}
```

Current situation:

* CMake compiles fine, both in terminal and with VSCode CMake extension.
* "Goto definition(ctrl+click)" works totally fine for every file.
* for files under MainModule, all looks corrupted. "Undefined symbol" error squiggles show everywhere with symbols from both MainModule and SubModuleA/B. `using namespace ABC` will get error hint "name must be namespace". Looks like all `#includes` not really working.
* for files under SubModuleA, "Undefined symbol" exists too, but if I open corresponding `.h` file which defines these symbols, squiggles disappear for a while. closing these `.h` files will make squiggles appear again.
* typing suggestions nearly unusable. suggestions barely listed there while typing symbols like function local variables, function parameter variables, types.

So, what would be the main fault? Is there any way to debugging/analyse Intellisense problems? now the VSCode just look like an annoying dumb text editor.

**Comment by sean-mcmanus** [comment]
I'll check if anyone on our team has time to look into this...

**Reply by bobbrow** [reply]
I see that the configuration provider is set to `ms-vscode.cmake-tools` so I'm going to transfer this over to the CMake Tools repo.  Can you run the `CMake: Log Diagnostics` command after opening the source files in the editor and share the output?  Also, please let us know how you include the SubModules in your CMakeLists.txt.  Is it with `add_directory`, `find_package`, or something else?  Thanks!

**Reply by epiciskandar** [reply]
This is a really complex project, `CMake: Log Diagnostics` gives me a bunch contents and hard to remove sensitive informations. Also the `CMakeLists.txt` refereced a global CMake script file which introducing dependencies by `target_link_libraries` and `target_include_directory`.

It feels hard to explain all situations, so I'm going to make a similar demo workspace and upload to github repo, we should investigate from there. This will be done in one week and repo URL will be applied here then. Thanks by the way!

**Reply by bobbrow** [reply]
Thanks for helping us reduce this into a testable demo project.  We'll be happy to investigate when it is ready.

**Reply by epiciskandar** [reply]
I made a demo workspace trying to reproduce this problem, but it works just fine. The problem might be anywhere else, so, I will try introducing more detailed differences until something touch the root cause.

**Comment by gcampbell-msft** [comment]
@epiciskandar Closing this discussion due to its age, but if you're still having issues and are able to provide a reproducible project that exemplifies any issue you're hitting, please create an issue and we'd be happy to look into it!
